java-checks/src/main/java/org/sonar/java/checks/SAMAnnotatedCheck.java

```diff
@@ -78,70 +83,52 @@ public void visitNode(Tree tree) {
   }
 
   private static boolean isAnnotated(ClassTree tree) {
-    for (AnnotationTree annotationTree : tree.modifiers().annotations()) {
-      Tree annotationType = annotationTree.annotationType();
-      if (annotationType.is(Tree.Kind.IDENTIFIER) && "FunctionalInterface".equals(((IdentifierTree) annotationType).name())) {
-        return true;
-      }
-    }
-    return false;
+    return tree.symbol().metadata().isAnnotatedWith("java.lang.FunctionalInterface");
   }
 
   private static boolean hasOneAbstractMethod(ClassTree classTree) {
-    TypeSymbol symbol = classTree.symbol();
-    if (symbol != null) {
-      List<Type> types = symbol.interfaces();
-      for (Type type : types) {
-        if (!type.symbol().memberSymbols().isEmpty()) {
-          return false;
-        }
-      }
+    TypeJavaSymbol symbol = (TypeJavaSymbol) classTree.symbol();
+    Set<ClassJavaType> superInterfaces = symbol.superTypes().stream().filter(type -> !type.is("java.lang.Object")).collect(Collectors.toSet());
+    if (superInterfaces.stream().anyMatch(SAMAnnotatedCheck::interfaceIsNotEmpty)) {
+      return false;
     }
-    int methods = 0;
-    for (Tree member : classTree.members()) {
-      boolean isMethod = member.is(Tree.Kind.METHOD);
-      if (!isMethod) {
-        return false;
-      }
-      if (isNotObjectMethod((MethodTree) member) && isNonStaticNonDefaultMethod(member)) {
-        methods++;
-      }
+    List<Tree> members = classTree.members();
+    if (members.stream().allMatch(tree -> tree.is(Tree.Kind.METHOD))) {
```

Comment:
This is not correct : A functional interfaces can have constants defined. or inner classes.